src/devices/src/legacy/acpi_device.rs

```diff
@@ -0,0 +1,166 @@
+            }
+        }
+    }
+}
+
+type Result<T> = result::Result<T, Error>;
+
+// AcpiDevice should emulate two ACPI registers required for a gracefull shutdown (using fixed hardware method):
+// X_PM1a_EVT_BLK (4 bytes) and X_PM1a_CNT_BLK (2 bytes)
+// https://uefi.org/specs/ACPI/6.4/05_ACPI_Software_Programming_Model/ACPI_Software_Programming_Model.html#fadt-format
+
+// To keep the implementation simple, both registers were implemented in the same
+// AcpiDevice; the bits of these two registers are saved into data field of the AcpiDevice
+// structure (in the least significant 6 bytes of the u64 variable) as it follows:
+
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// |                                                    data field of the AcpiDevice struct                                                  |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// | 0 ... 7       8        9 ... 19       24      25 ... 31 | 32  ...  41      42          43          44         45       46 47 | 48 .. 64 |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// |                        X_PM1a_EVT_BLK                   |                                  X_PM1a_CNT_BLK                    | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// | 0 ... 7       8        9 ... 19       24      25 ... 31 |  0  ...   9      10          11          12         13       14 15 | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+// |           PWRBTN_STS              PWRBTN_EN             |               SLP_TYPx    SLP_TYPx    SLP_TYPx    SLP_EN           | NOT USED |
+// -------------------------------------------------------------------------------------------------------------------------------------------
+
+// To trigger a shutdown, PWRBTN_STS and PWRBTN_EN (in X_PM1a_EVT_BLK) needs to be written as
+// ones and the interrupt should be raised. When guest goes to soft-off state, SLP_EN is written
+// as one and SLP_TYPx is written with the value from the _S5 object from the DSDT table (in
+// fact, when guest goes into power state x, then the value from the _Sx object is written into
+// SLP_TYPx bits located into X_PM1a_CNT_BLK)
+
+pub struct AcpiDevice {
+    /// Data Register
+    data: u64,
+    /// Interrupt to be triggered for OSPM to read ACPI registers
+    interrupt: EventFd,
+    /// CPU reset eventfd. We will set this event to stop Firecracker.
+    reset_evt: EventFd,
+}
+
+impl AcpiDevice {
```

Comment:
I think device unit tests should be fairly easy to add for this device. Right now there's no tests.